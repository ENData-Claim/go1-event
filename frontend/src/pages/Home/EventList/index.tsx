import React from 'react'

import { usePaginationFragment, useLazyLoadQuery } from 'react-relay/hooks'
import { graphql } from 'babel-plugin-relay/macro'
import { EventListPaginationQuery } from './__generated__/EventListPaginationQuery.graphql'
import { EventList_query$key } from './__generated__/EventList_query.graphql'
import { EventListQuery } from './__generated__/EventListQuery.graphql'

import { List, ListItem } from '@material-ui/core'
import { ErrorBoundary } from 'react-error-boundary'
import ErrorBoundaryWithRetry from 'components/ErrorBoundaryWithRetry'
import EventListItem from '../EventListItem'

type EventListPaginationProps = {
  query: EventList_query$key
}
const EventListPagination: React.FC<EventListPaginationProps> = props => {
  const { data, hasNext, isLoadingNext, loadNext } = usePaginationFragment<EventListPaginationQuery, EventList_query$key>(
    graphql`
      fragment EventList_query on Query
      @argumentDefinitions(
        count: { type: "Int", defaultValue: 5 }
        cursor: { type: "String" }
      )
      @refetchable(queryName: "EventListPaginationQuery")
      {
        eventConnection(first: $count, after: $cursor, where: $where)
        @connection(key: "EventList_eventConnection")
        {
          edges {
            cursor
            node {
              ...EventListItem_event
            }
          }
        }
      }
    `,
    props.query
  )

  return (
    <List>
      {data.eventConnection?.edges?.map(edge => {
        if(!edge?.node) return null

        return <EventListItem key={edge.cursor} event={edge.node} />
      })}
      {hasNext && (
        <ListItem
          button
          disabled={isLoadingNext}
          onClick={() => loadNext(5)}
        >
          Load more
        </ListItem>
      )}
    </List>
  )
}

const EventListLazyLoad: React.FC = () => {
  const query = useLazyLoadQuery<EventListQuery>(
    graphql`
      query EventListQuery($where: EventConnectionWhereInput) {
        ...EventList_query
      }
    `,
    {
      // where: {
      //   keywords: '',
      //   time: '',
      //   city: '',
      //   state: '',
      //   country: '',
      // }
    },
  )
    
  return <EventListPagination query={query} />
}

const EventList: React.FC = () => {
  return (
    <ErrorBoundary FallbackComponent={ErrorBoundaryWithRetry}>
      <React.Suspense fallback='Loading events...'>
        <EventListLazyLoad />
      </React.Suspense>
    </ErrorBoundary>
  )
}
export default EventList