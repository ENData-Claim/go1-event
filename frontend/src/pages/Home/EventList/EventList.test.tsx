import React from 'react'
import { render, cleanup, waitFor } from '@testing-library/react'
import { createMockEnvironment, MockPayloadGenerator } from 'relay-test-utils'
import { RelayEnvironmentProvider } from 'react-relay/hooks'

import EventList from '.'
import { BrowserRouter } from 'react-router-dom'
import { MuiPickersUtilsProvider } from '@material-ui/pickers'
import DateFnsUtils from '@date-io/date-fns'

afterEach(cleanup)

describe('<EventList />', () => {
  it('should reject query and catched by error boundary', async () => {
    const environment = createMockEnvironment()
    const { getByText } = render(
      <RelayEnvironmentProvider environment={environment}>
        <BrowserRouter>
          <EventList />
        </BrowserRouter>
      </RelayEnvironmentProvider>
    )

    environment.mock.rejectMostRecentOperation(operation => new Error(`Some error`))
    await waitFor(() => expect(getByText('Something went wrong:')).toBeTruthy())
  })
  it('should render loading then render EventList', async () => {
    const environment = createMockEnvironment()
    const { getByText, findByText, container } = render(
      <RelayEnvironmentProvider environment={environment}>
        <BrowserRouter>
          <MuiPickersUtilsProvider utils={DateFnsUtils}>
            <EventList />
          </MuiPickersUtilsProvider>
        </BrowserRouter>
      </RelayEnvironmentProvider>
    )

    expect(getByText('Loading events...')).toBeTruthy()

    environment.mock.resolveMostRecentOperation(operation => {
      return MockPayloadGenerator.generate(operation, {
        PageInfo() {
          return {
            "hasNextPage": true,
            "hasPreviousPage": false,
            "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
            "endCursor": "YXJyYXljb25uZWN0aW9uOjE="
          }
        },
        WorldEdge() {
          return [
            {
              "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
              "node": {
                "id": "RXZlbnQ6Y2tkbGRmZnc2MDAxOHJzeGFkN3UwYTg3Yg==",
                "title": "Place 1",
                "time": "2018-07-22T02:30:00.000Z",
                "image": "http://example.com/image.png",
                "location": {
                  "city": "Brisbane",
                  "state": "Queensland",
                  "country": "Australia"
                },
                "availableSeats": []
              }
            },
            {
              "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
              "node": {
                "id": "RXZlbnQ6Y2tkbGRmZzFwMDAyNHJzeGE2N3pjcWs3eg==",
                "title": "Place 2",
                "time": "2018-07-24T02:30:00.000Z",
                "image": "http://example.com/image.png",
                "location": {
                  "city": "Cairns",
                  "state": "Queensland",
                  "country": "Australia"
                },
                "availableSeats": [
                  {
                    "id": "U2VhdDpCMjk="
                  },
                  {
                    "id": "U2VhdDpXMjU="
                  }
                ]
              }
            }
          ]
        },
      })
    })

    await findByText(/Keywords/)
    expect(container).toMatchSnapshot()
  })
})