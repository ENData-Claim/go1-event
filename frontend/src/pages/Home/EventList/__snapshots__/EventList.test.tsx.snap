// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventList /> should render loading then render EventList 1`] = `
<div>
  <li
    class="MuiListItem-root MuiListItem-gutters"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
      >
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
          >
            Keywords
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Go
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
      >
        <div
          class="MuiFormControl-root MuiTextField-root"
          format="MMMM do"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink="true"
          >
            Time
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              readonly=""
              type="text"
              value="August 8th"
            />
          </div>
          
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
          >
            City
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              role="button"
              tabindex="0"
            >
              <span>
                â€‹
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </li>
  <ul
    class="MuiList-root MuiList-padding"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          &lt;mock-value-for-field-"title"&gt;
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary"
          >
            &lt;mock-value-for-field-"city"&gt;, &lt;mock-value-for-field-"state"&gt;, &lt;mock-value-for-field-"country"&gt;
          </p>
          <div>
            Invalid Date
          </div>
          1
           
          seat
           left
        </p>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      Load more
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </ul>
</div>
`;
